File: frontend/src/api/health.ts
import { apiClient } from './client';

export interface ApiErrorDetail {
  field?: string;
  message: string;
}

export interface ApiEnvelope<TData> {
  success: boolean;
  message: string;
  data: TData;
  errors: ApiErrorDetail[] | null;
}

export interface HealthData {
  status: 'healthy';
  service: string;
  version?: string;
}

export type HealthResponse = ApiEnvelope<HealthData>;

export const fetchHealth = async (): Promise<HealthData> => {
  const { data: response } = await apiClient.get<HealthResponse>('/health');

  if (!response.success || response.data.status !== 'healthy') {
    throw new Error(response.message || 'The backend returned an unhealthy status.');
  }

  return response.data;
};

File: backend/app/api/v1/endpoints/health.py
"""Service health endpoint."""

from typing import Annotated

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import Settings, get_settings
from app.database.session import get_db_session
from app.schemas.common import ApiResponse
from app.schemas.health import DependencyHealthData, HealthData
from app.services.health import HealthService, get_health_service

router = APIRouter()
Session = Annotated[AsyncSession, Depends(get_db_session)]
Configuration = Annotated[Settings, Depends(get_settings)]


@router.get(
    "",
    response_model=ApiResponse[HealthData],
    status_code=status.HTTP_200_OK,
    summary="Check API health",
)
async def health_check(
    settings: Configuration,
    service: Annotated[HealthService, Depends(get_health_service)],
) -> ApiResponse[HealthData]:
    """Return a lightweight liveness response without requiring the database."""
    return ApiResponse(
        success=True,
        message="Service is healthy.",
        data=service.get_health(version=settings.app_version),
        errors=None,
    )


@router.get(
    "/dependencies",
    response_model=ApiResponse[DependencyHealthData],
    status_code=status.HTTP_200_OK,
    summary="Check lightweight dependency readiness",
)
async def dependency_health_check(
    session: Session,
    settings: Configuration,
    service: Annotated[HealthService, Depends(get_health_service)],
) -> ApiResponse[DependencyHealthData]:
    """Report readiness without loading OCR, language, or similarity models."""
    return ApiResponse(
        success=True,
        message="Dependency readiness checked.",
        data=await service.get_dependency_health(
            session=session,
            settings=settings,
        ),
        errors=None,
    )
